File: suho/withjava/src/baekjoon/class3/B2630.java
package baekjoon.class3;

import java.io.*;
import java.util.StringTokenizer;

/*
플랫폼 : 백준
문제번호 : 2630
문제제목 : 색종이 만들기
난이도 : 실버 3
제한사항 : 1초/128MB
알고리즘 분류 : 분할 정복, 재귀

알고리즘 설명
1. 재귀로 푼다. -> step이 1이거나 색종이가 다 같은 색이면 종료
2. step을 /2 해가면서 분할 정복

채점 결과 : 236ms/17400KB
풀이 날짜 : 2021/09/20
*/

public class B2630 {
    static int[][] paper;
    static int n;
    static int white, blue;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        paper = new int[n][n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                paper[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        white = 0;
        blue = 0;
        dfs(0, 0, n);

        bw.write(white + "\n" + blue);
        bw.flush();
        bw.close();
        br.close();
    }

    static boolean isAllSameColor(int r, int c, int step) {
        int color = paper[r][c];
        for (int i = r; i < r + step; i++) {
            for (int j = c; j < c + step; j++) {
                if (paper[i][j] != color) {
                    return false;
                }
            }
        }
        return true;
    }

    static void dfs(int r, int c, int step) {
        if (step == 1) {
            if (paper[r][c] == 0) {
                white++;
            } else {
                blue++;
            }
            return;
        }

        if (isAllSameColor(r, c, step)) {
            if (paper[r][c] == 0) {
                white++;
            } else {
                blue++;
            }
            return;
        }

        for (int i = r; i < r + step; i += step / 2) {
            for (int j = c; j < c + step; j += step / 2) {
                dfs(i, j, step / 2);
            }
        }
    }
}

File: jaeseuk/kokotlin/src/main/kotlin/boj/recursion/B2630.kt
package boj.recursion

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

private val n = br.readLine().toInt()

private val paper = Array(n) { br.readLine().split(" ") }

private var white = 0
private var blue = 0

fun main() {
    solution()

    bw.append("$white\n")
    bw.append("$blue\n")
    bw.flush()
    bw.close()
}

private fun solution(startX: Int = 0, startY: Int = 0, endX: Int = n, endY: Int = n) {
    if (check(startX, startY, endX, endY)) return

    val nx = (startX + endX) / 2
    val ny = (startY + endY) / 2

    solution(startX, startY, nx, ny)
    solution(startX, ny, nx, endY)
    solution(nx, startY, endX, ny)
    solution(nx, ny, endX, endY)
}

private fun check(startX: Int, startY: Int, endX: Int, endY: Int): Boolean {
    val color = paper[startX][startY]

    for (i in startX until endX) {
        for (j in startY  until endY) {
            if (paper[i][j] != color) return false
        }
    }

    if (color == "1") blue++
    else white++

    return true
}